Settings ZMS backend optimizations

rghaddab opened the pull request:
These are series of optimization for the new backend + some commits to syncronize ZMS to the upstream version

rghaddab committed:
[nrf fromtree] settings: ZMS: add a backend for ZMS (Zephyr Memory Storage)

This adds the initial backend support for the ZMS storage system.

Signed-off-by: Riadh Ghaddab <rghaddab@baylibre.com>
(cherry picked from commit ef4e8dd5c3f8186ab62d28d92bd3f31ddefade61)

omkar3141 committed:
[nrf fromtree] tests: settings: Performance test for settings

This adds a on target performance test for Settings SS. Using this test
performance of the Setting SS + NVS/ZMS backend can be benchmarked.

The test repeatedly write 128 settings entries. Each setting entry
has a size of 4 bytes, and path length of 16 bytes (excluding the
null-terminator).

The test has two variants, with and without Bluetooth scan running.
This is useful to benchmark performance along with some component
of BT subsystem activated.

The test could be enhanced in future to include or create combinations
of different functionalities running, when agreesive store operations
are happening.

Signed-off-by: Omkar Kulkarni <omkar.kulkarni@nordicsemi.no>
(cherry picked from commit 97166a5c1115c34544df5c031f4ccaba22a6b403)

rghaddab committed:
[nrf fromtree] fs: zms: multiple style fixes from previous PR review

This resolves some addressed comments in this PR zephyrproject-rtos#77930
as well as this PR zephyrproject-rtos#80407

Signed-off-by: Riadh Ghaddab <rghaddab@baylibre.com>
(cherry picked from commit 5f7cda5e0608bd027f759736aef78bf6f03070d0)

rghaddab committed:
Revert "[nrf noup] settings: zms: use dedicated lookup cache hash function"

This reverts commit 14d166ffe235587818ce069c85c5b819c0b170a5.

Signed-off-by: Riadh Ghaddab <rghaddab@baylibre.com>

rghaddab committed:
Revert "[nrf fromlist] settings: ZMS: add a backend for ZMS (Zephyr Memory Storage)"

This reverts commit 88e9f91ffa5b30f3b20c8fed5dbcd5cc88b1d32e.

Signed-off-by: Riadh Ghaddab <rghaddab@baylibre.com>

rghaddab committed:
[nrf fromtree] settings: zms: fix some bugs related to the name's ID

To avoid collisions between IDs used by settings and IDs used directly
by subsystems using ZMS API, the MSB is always set to 1 for Setting's
name ID written to ZMS backend

Add as well a recovery path if the hash linked list is broken.

Signed-off-by: Riadh Ghaddab <rghaddab@baylibre.com>
(cherry picked from commit 333faddd43ff4958bbcb441fd3aa4e7e2f7cf371)

rghaddab committed:
[nrf fromtree] doc: settings: add reference to ZMS backend

Add reference to the new settings backend ZMS and add more details about
how it works.

Signed-off-by: Riadh Ghaddab <rghaddab@baylibre.com>
(cherry picked from commit f3630157e03fa322787f264b1047f91da5c2af8c)

aescolar committed:
[nrf fromtree] samples/subsys/fs/zms: run on native_sim instead of native_posix

native_posix is deprecated and will be removed in 4.2
Run this sample in native_sim instead.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>
(cherry picked from commit 4a91a55ae24f4afbd4893661637f263560636c0d)

aescolar committed:
[nrf fromtree] samples/subsys/fs/zms: Allow it to run and add check

This sample could not run on any target as it depedend on zms
which is set by no board.
But it seems to run just fine in both allowed platforms,
so let's add a trivial twister check (so it actually passes instead
of waiting for ever), and remove that `depends on`.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>
(cherry picked from commit cc2d31f4da1446c54baef7688fb41761658e44b9)

butok committed:
[nrf fromtree] sample: zms: add ZMS sample Kconfig

- adds ability to change the ZMS sample MAX_ITERATIONS
  and DELETE_ITERATION parameters via Kconfig without manually
  modifying the source code.
- adds ability to reduce flash memory wear on real devices
  by reducing the number of write iterations.

Signed-off-by: Andrej Butok <andrey.butok@nxp.com>
(cherry picked from commit c887d5e235478cdf59381759a081c15f68a12869)

butok committed:
[nrf fromtree] tests: zms: allow to build & run on real platforms

- Allows to build and run ZMS tests on real platforms.
- Enables ZMS tests designed for a flash-simulator only when
  built for qemu_x86 or native_sim targets.

Signed-off-by: Andrej Butok <andrey.butok@nxp.com>
(cherry picked from commit b826d4663226909d49ea85bab707ee31dcdd86f8)

rghaddab committed:
[nrf fromlist] settings: zms: add option to disable updating linked list

When deleting a settings entry the linked list is updated by removing
the deleted node. This operation is time consuming and add some delays.
For applications that use almost the same set of Setting entries, add an
option to make this operation faster at a cost of having some empty
nodes in the linked list.
These empty nodes can be used later when the deleted entry is written
again.
Each empty node occupies 16B of storage.

Upstream PR #: 87792

Signed-off-by: Riadh Ghaddab <rghaddab@baylibre.com>
(cherry picked from commit 02ab54e077e090e4c9364e4d4954a8c3508e513e)

butok committed:
[nrf fromtree] samples: zms: fix the loop_cnt iteration 0 check

- fix unwanted "Error: Something went wrong at iteration 0".
- the loop_cnt check has no sense for iteration 0,
  comparing with wrong -1.

Signed-off-by: Andrej Butok <andrey.butok@nxp.com>
(cherry picked from commit 2766c788230f54bd30d3b4f1d95af4e87f0c3de8)

butok committed:
[nrf fromtree] samples: zms: fix 9999 messages dropped

- Sets CONFIG_LOG_BLOCK_IN_THREAD for the zms sample.
- Fixes "--- 9999 messages dropped ---".

Signed-off-by: Andrej Butok <andrey.butok@nxp.com>
(cherry picked from commit ac7790e452ddae08fc4d9c72d7f74ba50c7e1dec)

rghaddab committed:
[nrf fromlist] zms: remove non needed lookup cache before writing

When the CONFIG_ZMS_NO_DOUBLE_WRITE is not enabled there is no need to
search in the cache for matching ID

Upstream PR #: 87792

Signed-off-by: Riadh Ghaddab <rghaddab@baylibre.com>
(cherry picked from commit 546ed2dbe817794bedb16bae0c81fd4af22675ea)

rghaddab committed:
[nrf fromlist] zms: optimize cache for settings subsystem

Settings subsystem is storing the name ID and the linked list node ID
with only one bit difference at BIT(0).
Settings subsystem is also storing the name ID and the data ID in two
different ZMS entries at an exact offset of ZMS_DATA_ID_OFFSET.
Using the regular lookup function could result in many cache misses.

Therefore, to assure the least number of collisions in the lookup cache,
the BIT(0) of the hash indicates whether the given ZMS ID represents a
linked list entry or not, the BIT(1) indicates whether the ZMS ID is a
name or data and the remaining bits of the hash are set to a truncated
part of the original hash generated by Settings.

Upstream PR #: 87792

Signed-off-by: Riadh Ghaddab <rghaddab@baylibre.com>
(cherry picked from commit ad2581fede5f6efa570e3e8157ff37290d2dbe21)

rghaddab committed:
[nrf fromlist] settings: zms: load only subtree in the argument

If the subtree argument is not NULL in the settings_load function, load
only the setting entry that corresponds to that subtree.
If the subtree argument is NULL or it is not found in the storage, load
all the existing entries in the storage.

Upstream PR #: 87792

Signed-off-by: Riadh Ghaddab <rghaddab@baylibre.com>
(cherry picked from commit 86132feb5c59e8a071d9782c5482ab346cfbf6f3)

rghaddab committed:
[nrf fromlist] settings: zms: add cache for linked list hash

Increase the load performance by adding an optional cache for the linked
list hashes.
This is used only when the settings_load is called with NULL parameter
and we need to load all Settings that exist in the persistent storage.
Cache is enabled using SETTINGS_ZMS_LL_CACHE and the size of the cache
is set using SETTINGS_ZMS_LL_CACHE_SIZE.
Each cache entry will add 8 bytes of RAM usage.

Upstream PR #: 87792

Signed-off-by: Riadh Ghaddab <rghaddab@baylibre.com>
(cherry picked from commit 50c099243557a49b53d793c839299e0a55370f42)

Damian-Nordic commented on subsys/settings/src/settings_zms.c:
```diff
@@ -207,7 +207,41 @@ static int settings_zms_load(struct settings_store *cs, const struct settings_lo
 
+			ret = settings_call_set_handler(arg->subtree, rc2, settings_zms_read_fn,
+							&read_fn_arg, (void *)arg);
+			return ret;
```
Unfortunately, you cannot return here. If `subtree == "abc"`, the code is expected to return not only "abc" but also "abc/def" etc. You can only break if `ret != 0`.

rghaddab replied:
@Damian-Nordic you are right I will change that.
I actually didn't see any application using this feature of settings where you can load subtree "abc" and it will load abc/*

rghaddab committed:
[nrf fromlist] settings: add an API function to load only one settings entry

Add a a function settings_load_one that allows to load only one settings
entry instead of a complete subtree.
Add as well its implementation for ZMS backend.
This will allow a faster return if the Settings entry doesn't exist in
the persistent storage.

Upstream PR #: 87792

Signed-off-by: Riadh Ghaddab <rghaddab@baylibre.com>
(cherry picked from commit 98e544b4b838cd96012bb4bd5adfb584e67a52f1)

rghaddab committed:
[nrf fromlist] settings: zms: add more robustness to the save function

When a power off happens after writing the settings name and before
writing the linked list node we cannot write the settings name again in
the future.
Fix this by writing the linked list node before writing the name.
When loading all settings, we already delete linked list node that do
not have any name or value written.
Adds as well a recover path if a power down happens in the middle of
unlinking an LL node after a delete.

Upstream PR #: 87792

Signed-off-by: Riadh Ghaddab <rghaddab@baylibre.com>
(cherry picked from commit bf4d15316b1a3bd9239ddbe179b11006eea1e3b1)

rghaddab committed:
[nrf fromlist] settings: zms: recover linked list if broken

When the linked list is broken, recover it instead of reinitializing it.

Upstream PR #: 87792

Signed-off-by: Riadh Ghaddab <rghaddab@baylibre.com>
(cherry picked from commit 057c1d65fa5822d9145330c87183d0b652707ab2)

rghaddab committed:
[nrf fromlist] settings: zms: use the safe function strnlen instead of strlen

if the provided name in argument is not null this could lead to un
undefined behavior.
Use strnlen to make this safe

Upstream PR #: 87792

Signed-off-by: Riadh Ghaddab <rghaddab@baylibre.com>
(cherry picked from commit e9ac2ef95bd34a8c4756e46e8f2ddf66e0964dbc)

rghaddab committed:
[nrf fromlist] tests: settings: add a functional test for settings_load_one

Add a test for the new API settings_load_one that loads only one path
from the persistent storage.

Upstream PR #: 87792

Signed-off-by: Riadh Ghaddab <rghaddab@baylibre.com>
(cherry picked from commit 168ae245bdce756fd5db22dd391469ad17eeb9b2)

rghaddab committed:
[nrf fromlist] tests: settings: add functional test for ZMS

Add the functional test for the new backend of ZMS

Upstream PR #: 87792

Signed-off-by: Riadh Ghaddab <rghaddab@baylibre.com>
(cherry picked from commit 18f7736ce3c73f6f10e037aa627293104dc55b9d)

rghaddab committed:
[nrf fromlist] settings: add new API function settings_get_val_len()

Add a function to get the value's length of a Key. If it doesn't exist
returns 0.
Add ZMS implementation for csi_get_val_len() and a default implementation
for the other storage systems.
Add some functional tests to verify it.

Upstream PR #: 87792

Signed-off-by: Riadh Ghaddab <rghaddab@baylibre.com>
(cherry picked from commit 5a730f71f49e87fb4da35120826e8d45bc62fef6)

rghaddab committed:
[nrf fromlist] doc: settings: new API functions

Add references to the new API functions that were added
"csi_load_one" and "csi_get_val_len"

Upstream PR #: 87792

Signed-off-by: Riadh Ghaddab <rghaddab@baylibre.com>
(cherry picked from commit b26ed3a3ec5dae5017ecd4e40164e3b38ca88235)

omkar3141 reviewed (dismissed):
Could you check sonar cloud issue and see if you can resolve as many as possible if implementation is settled?

I haven't re-read the full PR yet.

rghaddab committed:
[nrf fromlist] settings: zms: code style clean up

Clean some parts of the code and refactor it to avoid multiple nested
conditions.

Upstream PR #: 87792

Signed-off-by: Riadh Ghaddab <rghaddab@baylibre.com>
(cherry picked from commit 1699785ff3fd7d7e260133898729afff634eea73)

rghaddab commented:
> Could you check sonar cloud issue and see if you can resolve as many as possible if implementation is settled?
> 
> I haven't re-read the full PR yet.

I resolved as many Sonar Cloud Issues as possible.

carlescufi commented:
@omkar3141 @Damian-Nordic please take another look
